mono/metadata/unity-liveness.c

```diff
@@ -1,100 +1,150 @@
+	}
+	*block_array->current_block->next_item++ = value;
 }
 
-void array_push_back(custom_growable_array* array, gpointer value)
+gpointer block_array_pop_back(custom_growable_block_array *block_array)
 {
-	g_assert(!array_is_full(array));
-	array->pdata[array->len] = value;
-	array->len++;
+	if (block_array->current_block->next_item == block_array->current_block->p_data) {
+		if (block_array->current_block->prev_block == NULL)
+			return NULL;
+		block_array->current_block = block_array->current_block->prev_block;
+		block_array->current_block->next_item = block_array->current_block->p_data + k_array_elements_per_block;
+	}
+	return *--block_array->current_block->next_item;
 }
 
-gpointer array_pop_back(custom_growable_array* array)
+void block_array_reset_iterator(custom_growable_block_array *array)
 {
-	array->len--;
-	return array->pdata[array->len];
+	array->iterator->current_block = array->first_block;
+	array->iterator->current_position = array->first_block->p_data;
 }
 
-void array_clear(custom_growable_array* array)
+gpointer block_array_next(custom_growable_block_array *block_array)
 {
-	array->len = 0;
+	custom_block_array_iterator *iterator = block_array->iterator;
+	if (iterator->current_position != iterator->current_block->next_item)
+		return *iterator->current_position++;
+	if (iterator->current_block->next_block == NULL)
+		return NULL;
+	iterator->current_block = iterator->current_block->next_block;
+	iterator->current_position = iterator->current_block->p_data;
+	if (iterator->current_position == iterator->current_block->next_item)
+		return NULL;
+	return *iterator->current_position++;
 }
 
-void array_grow(custom_growable_array* array)
+void block_array_clear(custom_growable_block_array *block_array)
 {
-	int oldlen = array->len;
-	g_ptr_array_set_size(array, g_ptr_array_capacity(array)*2);
-	array->len = oldlen;
+	custom_array_block *block = block_array->first_block;
+	while (block != NULL) {
+		block->next_item = block->p_data;
+		block = block->next_block;
+	}
 }
 
-typedef void (*register_object_callback)(gpointer* arr, int size, void* callback_userdata);
-typedef void (*WorldStateChanged)();
-struct _LivenessState
+void block_array_destroy(custom_growable_block_array *block_array, LivenessState *state)
 {
-	gint                first_index_in_all_objects;
-	custom_growable_array* all_objects;
-
-	MonoClass*          filter;
-
-	custom_growable_array* process_array;
-	guint               initial_alloc_count;
-
-	void*               callback_userdata;
-
-	register_object_callback filter_callback;
-	WorldStateChanged        onWorldStartCallback;
-	WorldStateChanged        onWorldStopCallback;
-	guint               traverse_depth; // track recursion. Prevent stack overflow by limiting recurion
-};
+	custom_array_block *block = block_array->first_block;
+	while (block != NULL) {
+		void *data_block = block;
+		block = block->next_block;
+		state->reallocateArray(data_block, 0, state->callback_userdata);
```

Comment:
The reallocate of size 0 indicates to "free" the block?

Reply:
yes. I used reallocate for all 3 operations to start with, but ended up with not using actual reallocate. Only allocate and free. Allcoate using NULL ptr and free using 0 size